File: shuttle-api/src/ShuttleApi.Domain/Vehicles/VehicleInspectionRecord.cs
using ShuttleApi.Domain.Common;

namespace ShuttleApi.Domain.Vehicles;

public sealed class VehicleInspectionRecord : Entity<Guid>
{
    public Guid VehicleId { get; private set; }
    public InspectionType InspectionType { get; private set; }

    public DateTime InspectedAt { get; private set; }
    public DateTime? ExpiresAt { get; private set; }

    public string? InspectorName { get; private set; }
    public string? InspectionFacility { get; private set; }
    public string? CertificateNumber { get; private set; }

    public InspectionResult InspectionResult { get; private set; }

    public string? DeficienciesNotes { get; private set; }
    public string? CorrectiveActionNotes { get; private set; }

    public decimal? CostDollars { get; private set; }

    public DateTime CreatedAt { get; private set; }

    public bool IsExpiringSoon =>
        ExpiresAt.HasValue && ExpiresAt.Value <= DateTime.UtcNow.AddDays(60);

    private VehicleInspectionRecord() { }

    public static VehicleInspectionRecord Create(
        Guid vehicleId,
        InspectionType inspectionType,
        DateTime inspectedAt,
        DateTime? expiresAt,
        string? inspectorName,
        string? inspectionFacility,
        string? certificateNumber,
        InspectionResult inspectionResult,
        string? deficienciesNotes,
        string? correctiveActionNotes,
        decimal? costDollars)
    {
        return new VehicleInspectionRecord
        {
            Id = Guid.NewGuid(),
            VehicleId = vehicleId,
            InspectionType = inspectionType,
            InspectedAt = inspectedAt,
            ExpiresAt = expiresAt,
            InspectorName = inspectorName,
            InspectionFacility = inspectionFacility,
            CertificateNumber = certificateNumber,
            InspectionResult = inspectionResult,
            DeficienciesNotes = deficienciesNotes,
            CorrectiveActionNotes = correctiveActionNotes,
            CostDollars = costDollars,
            CreatedAt = DateTime.UtcNow
        };
    }

    public void Update(
        InspectionType inspectionType,
        DateTime inspectedAt,
        DateTime? expiresAt,
        string? inspectorName,
        string? inspectionFacility,
        string? certificateNumber,
        InspectionResult inspectionResult,
        string? deficienciesNotes,
        string? correctiveActionNotes,
        decimal? costDollars)
    {
        InspectionType = inspectionType;
        InspectedAt = inspectedAt;
        ExpiresAt = expiresAt;
        InspectorName = inspectorName;
        InspectionFacility = inspectionFacility;
        CertificateNumber = certificateNumber;
        InspectionResult = inspectionResult;
        DeficienciesNotes = deficienciesNotes;
        CorrectiveActionNotes = correctiveActionNotes;
        CostDollars = costDollars;
    }
}

File: shuttle_tablet/lib/features/vehicles/domain/entities/vehicle_inspection_record.dart
import 'package:equatable/equatable.dart';

class VehicleInspectionRecord extends Equatable {
  final String id;
  final String vehicleId;
  final String inspectionType;
  final DateTime inspectedAt;
  final DateTime? expiresAt;
  final String? inspectorName;
  final String? inspectionFacility;
  final String? certificateNumber;
  final String inspectionResult;
  final String? deficienciesNotes;
  final String? correctiveActionNotes;
  final double? costDollars;
  final DateTime createdAt;
  final bool isExpiringSoon;

  const VehicleInspectionRecord({
    required this.id,
    required this.vehicleId,
    required this.inspectionType,
    required this.inspectedAt,
    this.expiresAt,
    this.inspectorName,
    this.inspectionFacility,
    this.certificateNumber,
    required this.inspectionResult,
    this.deficienciesNotes,
    this.correctiveActionNotes,
    this.costDollars,
    required this.createdAt,
    this.isExpiringSoon = false,
  });

  bool get isPassed => inspectionResult.toLowerCase() == 'pass';
  bool get isFailed => inspectionResult.toLowerCase() == 'fail';

  @override
  List<Object?> get props => [
        id,
        vehicleId,
        inspectionType,
        inspectedAt,
        expiresAt,
        inspectorName,
        inspectionFacility,
        certificateNumber,
        inspectionResult,
        deficienciesNotes,
        correctiveActionNotes,
        costDollars,
        createdAt,
        isExpiringSoon,
      ];
}
